// @neo4j/graphql/dist/translate/cypher-builder/variables/NodeRef.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NamedNode = exports.NodeRef = void 0;
const Pattern_1 = require("../Pattern");
const utils_1 = require("../utils/utils");
const Variable_1 = require("./Variable");
/** Represents a Node reference */
class NodeRef extends Variable_1.Variable {
    constructor(options) {
        super("this");
        this.labels = options.labels || [];
    }
    getLabelsString() {
        const escapedLabels = this.labels.map(utils_1.escapeLabel);
        if (escapedLabels.length === 0)
            return "";
        return `:${escapedLabels.join(":")}`;
    }
    /** Creates a new Pattern from this node */
    pattern(options = {}) {
        return new Pattern_1.Pattern(this, options);
    }
}
exports.NodeRef = NodeRef;
class NamedNode extends NodeRef {
    constructor(id, options) {
        super(options || {});
        this.id = id;
    }
}
exports.NamedNode = NamedNode;
