// @neo4j/graphql/dist/translate/where/property-operations/create-comparison-operation.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createComparisonOperation = void 0;
const CypherBuilder = __importStar(require("../../cypher-builder/CypherBuilder"));
const create_point_comparison_operation_1 = require("./create-point-comparison-operation");
/** Translates an atomic comparison operation (e.g. "this0 <= $param0") */
function createComparisonOperation({ operator, propertyRefOrCoalesce, param, durationField, pointField, neo4jDatabaseInfo, }) {
    if (pointField) {
        return (0, create_point_comparison_operation_1.createPointComparisonOperation)({
            operator,
            propertyRefOrCoalesce,
            param,
            pointField,
            neo4jDatabaseInfo,
        });
    }
    // Comparison operations requires adding dates to durations
    // See https://neo4j.com/developer/cypher/dates-datetimes-durations/#comparing-filtering-values
    if (durationField && operator) {
        return createDurationOperation({ operator, property: propertyRefOrCoalesce, param });
    }
    return createBaseOperation({
        operator: operator || "EQ",
        property: propertyRefOrCoalesce,
        param,
    });
}
exports.createComparisonOperation = createComparisonOperation;
function createDurationOperation({ operator, property, param, }) {
    const variable = CypherBuilder.plus(CypherBuilder.datetime(), param);
    const propertyRef = CypherBuilder.plus(CypherBuilder.datetime(), property);
    return createBaseOperation({
        operator,
        property: propertyRef,
        param: variable,
    });
}
function createBaseOperation({ operator, property, param, }) {
    switch (operator) {
        case "LT":
            return CypherBuilder.lt(property, param);
        case "LTE":
            return CypherBuilder.lte(property, param);
        case "GT":
            return CypherBuilder.gt(property, param);
        case "GTE":
            return CypherBuilder.gte(property, param);
        case "ENDS_WITH":
        case "NOT_ENDS_WITH":
            return CypherBuilder.endsWith(property, param);
        case "STARTS_WITH":
        case "NOT_STARTS_WITH":
            return CypherBuilder.startsWith(property, param);
        case "MATCHES":
            return CypherBuilder.matches(property, param);
        case "CONTAINS":
        case "NOT_CONTAINS":
            return CypherBuilder.contains(property, param);
        case "IN":
        case "NOT_IN":
            return CypherBuilder.in(property, param);
        case "INCLUDES":
        case "NOT_INCLUDES":
            return CypherBuilder.in(param, property);
        case "EQ":
        case "NOT":
            return CypherBuilder.eq(property, param);
        default:
            throw new Error(`Invalid operator ${operator}`);
    }
}
