// @neo4j/graphql/dist/translate/translate-top-level-cypher.js
"use strict";
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.translateTopLevelCypher = void 0;
const graphql_1 = require("graphql");
const create_projection_and_params_1 = __importDefault(require("./create-projection-and-params"));
const create_auth_and_params_1 = require("./create-auth-and-params");
const constants_1 = require("../constants");
const create_connection_and_params_1 = __importDefault(require("./connection/create-connection-and-params"));
const CypherBuilder = __importStar(require("./cypher-builder/CypherBuilder"));
const get_neo4j_resolve_tree_1 = __importDefault(require("../utils/get-neo4j-resolve-tree"));
const create_auth_param_1 = __importDefault(require("./create-auth-param"));
const utils_1 = require("./cypher-builder/utils/utils");
function translateTopLevelCypher({ context, info, field, args, type, statement, }) {
    context.resolveTree = (0, get_neo4j_resolve_tree_1.default)(info);
    const { resolveTree } = context;
    let params = { ...args, auth: (0, create_auth_param_1.default)({ context }), cypherParams: context.cypherParams };
    const cypherStrs = [];
    const preAuth = (0, create_auth_and_params_1.createAuthAndParams)({ entity: field, context });
    if (preAuth[0]) {
        params = { ...params, ...preAuth[1] };
        cypherStrs.push(`CALL apoc.util.validate(NOT (${preAuth[0]}), "${constants_1.AUTH_FORBIDDEN_ERROR}", [0])`);
    }
    let projectionStr = "";
    const projectionAuthStrs = [];
    const projectionSubqueries = [];
    const connectionProjectionStrs = [];
    const referenceNode = context.nodes.find((x) => x.name === field.typeMeta.name);
    if (referenceNode) {
        const recurse = (0, create_projection_and_params_1.default)({
            resolveTree,
            node: referenceNode,
            context,
            varName: `this`,
        });
        const { projection: str, params: p, meta, subqueries } = recurse;
        projectionStr = str;
        projectionSubqueries.push(...subqueries);
        params = { ...params, ...p };
        if (meta.authValidateStrs?.length) {
            projectionAuthStrs.push(...projectionAuthStrs, meta.authValidateStrs.join(" AND "));
        }
        if (meta.connectionFields?.length) {
            meta.connectionFields.forEach((connectionResolveTree) => {
                const connectionField = referenceNode.connectionFields.find((x) => x.fieldName === connectionResolveTree.name);
                const nestedConnection = (0, create_connection_and_params_1.default)({
                    resolveTree: connectionResolveTree,
                    field: connectionField,
                    context,
                    nodeVariable: "this",
                });
                const [nestedStr, nestedP] = nestedConnection;
                connectionProjectionStrs.push(nestedStr);
                params = { ...params, ...nestedP };
            });
        }
    }
    const unionWhere = [];
    const graphqlType = context.schema.getType(field.typeMeta.name);
    const referenceUnion = graphqlType instanceof graphql_1.GraphQLUnionType ? graphqlType.astNode : undefined;
    if (referenceUnion) {
        const headStrs = [];
        const referencedNodes = referenceUnion.types
            ?.map((u) => context.nodes.find((n) => n.name === u.name.value))
            ?.filter((b) => b !== undefined)
            ?.filter((n) => Object.keys(resolveTree.fieldsByTypeName).includes(n?.name ?? "")) || [];
        referencedNodes.forEach((node) => {
            if (node) {
                const labelsStatements = node.getLabels(context).map((label) => `"${label}" IN labels(this)`);
                unionWhere.push(`(${labelsStatements.join("AND")})`);
                const innerHeadStr = [`[ this IN [this] WHERE (${labelsStatements.join(" AND ")})`];
                if (resolveTree.fieldsByTypeName[node.name]) {
                    const { projection: str, params: p, meta, subqueries, } = (0, create_projection_and_params_1.default)({
                        resolveTree,
                        node,
                        context,
                        varName: "this",
                    });
                    projectionSubqueries.push(...subqueries);
                    innerHeadStr.push([`| this { __resolveType: "${node.name}", `, ...str.replace("{", "").split("")].join(""));
                    params = { ...params, ...p };
                    if (meta.authValidateStrs?.length) {
                        projectionAuthStrs.push(meta.authValidateStrs.join(" AND "));
                    }
                }
                else {
                    innerHeadStr.push(`| this { __resolveType: "${node.name}" } `);
                }
                innerHeadStr.push(`]`);
                headStrs.push(innerHeadStr.join(" "));
            }
        });
        projectionStr = `${headStrs.join(" + ")}`;
    }
    const initApocParamsStrs = ["auth: $auth", ...(context.cypherParams ? ["cypherParams: $cypherParams"] : [])];
    // Null default argument values are not passed into the resolve tree therefore these are not being passed to
    // `apocParams` below causing a runtime error when executing.
    const nullArgumentValues = field.arguments.reduce((res, argument) => ({
        ...res,
        ...{ [argument.name.value]: null },
    }), {});
    const apocParams = Object.entries({ ...nullArgumentValues, ...resolveTree.args }).reduce((result, entry) => ({
        strs: [...result.strs, `${entry[0]}: $${entry[0]}`],
        params: { ...result.params, [entry[0]]: entry[1] },
    }), { strs: initApocParamsStrs, params });
    params = { ...params, ...apocParams.params };
    const apocParamsStr = `{${apocParams.strs.length ? `${apocParams.strs.join(", ")}` : ""}}`;
    if (type === "Query") {
        const isArray = field.typeMeta.array;
        const expectMultipleValues = !field.isScalar && !field.isEnum && isArray;
        if (expectMultipleValues) {
            cypherStrs.push(`WITH apoc.cypher.runFirstColumnMany("${statement}", ${apocParamsStr}) as x`);
        }
        else {
            cypherStrs.push(`WITH apoc.cypher.runFirstColumnSingle("${statement}", ${apocParamsStr}) as x`);
        }
        cypherStrs.push("UNWIND x as this\nWITH this");
    }
    else {
        cypherStrs.push(`
            CALL apoc.cypher.doIt("${statement}", ${apocParamsStr}) YIELD value
            WITH apoc.map.values(value, [keys(value)[0]])[0] AS this
        `);
    }
    if (unionWhere.length) {
        cypherStrs.push(`WHERE ${unionWhere.join(" OR ")}`);
    }
    if (projectionAuthStrs.length) {
        cypherStrs.push(`WHERE apoc.util.validatePredicate(NOT (${projectionAuthStrs.join(" AND ")}), "${constants_1.AUTH_FORBIDDEN_ERROR}", [0])`);
    }
    cypherStrs.push(connectionProjectionStrs.join("\n"));
    const projectionSubquery = CypherBuilder.concat(...projectionSubqueries);
    return new CypherBuilder.RawCypher((env) => {
        const subqueriesStr = (0, utils_1.compileCypherIfExists)(projectionSubquery, env);
        if (subqueriesStr)
            cypherStrs.push(subqueriesStr);
        if (field.isScalar || field.isEnum) {
            cypherStrs.push(`RETURN this`);
        }
        else if (referenceUnion) {
            cypherStrs.push(`RETURN head( ${projectionStr} ) AS this`);
        }
        else {
            cypherStrs.push(`RETURN this ${projectionStr} AS this`);
        }
        return [cypherStrs.join("\n"), params];
    }).build();
}
exports.translateTopLevelCypher = translateTopLevelCypher;
